File: packages/python/alexandria_sdk/publish.py
"""Publish a packed `.atool` / `.aagent` archive to an Alexandria registry.

Mirrors the TypeScript SDK ``publish()``: re-verifies the archive (hashes +
schema) then POSTs a multipart body to ``{registry}/v1/submit`` — the missing
consumer-side half of the registry loop (``alexandria install`` pulls; nothing
pushed until now).

The multipart body mirrors the registry's ``handleSubmit`` contract exactly:

- ``artifact_type`` — derived from the manifest kind (mcp|atool|aagent),
  overridable for amodel sub-variants.
- ``tarball`` — the packed archive bytes.
"""

from __future__ import annotations

import json
import mimetypes
import os
import urllib.error
import urllib.request
import uuid
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Callable

from .pack import verify

# Injectable transport seam so tests can mock the HTTP call without a live
# server. A transport takes (url, body, headers) and returns (status, text).
Transport = Callable[[str, bytes, dict[str, str]], "tuple[int, str]"]


@dataclass
class PublishResult:
    status: int
    ok: bool
    body: Any
    name: str
    version: str
    artifact_type: str


def _build_multipart(artifact_type: str, filename: str, tarball: bytes) -> tuple[bytes, str]:
    """Assemble a multipart/form-data body with an ``artifact_type`` field and a
    ``tarball`` file part. Returns (body, content_type)."""
    boundary = f"----alexsdk{uuid.uuid4().hex}"
    ctype = mimetypes.guess_type(filename)[0] or "application/gzip"
    parts: list[bytes] = []

    parts.append(f"--{boundary}\r\n".encode())
    parts.append(b'Content-Disposition: form-data; name="artifact_type"\r\n\r\n')
    parts.append(artifact_type.encode() + b"\r\n")

    parts.append(f"--{boundary}\r\n".encode())
    parts.append(
        f'Content-Disposition: form-data; name="tarball"; filename="{filename}"\r\n'.encode()
    )
    parts.append(f"Content-Type: {ctype}\r\n\r\n".encode())
    parts.append(tarball)
    parts.append(b"\r\n")

    parts.append(f"--{boundary}--\r\n".encode())
    return b"".join(parts), f"multipart/form-data; boundary={boundary}"


def _default_transport(url: str, body: bytes, headers: dict[str, str]) -> tuple[int, str]:
    req = urllib.request.Request(url, data=body, headers=headers, method="POST")
    try:
        with urllib.request.urlopen(req) as resp:  # noqa: S310 (trusted registry URL)
            return resp.status, resp.read().decode("utf-8", errors="replace")
    except urllib.error.HTTPError as e:
        return e.code, e.read().decode("utf-8", errors="replace")


def publish(
    pkg_path: str | Path,
    registry: str,
    token: str | None = None,
    artifact_type: str | None = None,
    *,
    transport: Transport | None = None,
) -> PublishResult:
    """Publish ``pkg_path`` to ``{registry}/v1/submit``.

    The archive is re-verified before it ships — never publish an archive the
    local runtime would itself reject. Pass ``transport`` to inject a custom
    HTTP client (used by tests to mock the call).
    """
    manifest = verify(pkg_path)
    resolved_type = artifact_type or str(manifest["kind"])

    tarball = Path(pkg_path).read_bytes()
    body, content_type = _build_multipart(resolved_type, os.path.basename(str(pkg_path)), tarball)

    base = registry.rstrip("/")
    url = f"{base}/v1/submit"
    headers: dict[str, str] = {
        "Content-Type": content_type,
        "Content-Length": str(len(body)),
    }
    if token:
        headers["Authorization"] = f"Bearer {token}"

    send = transport or _default_transport
    status, text = send(url, body, headers)

    try:
        parsed: Any = json.loads(text)
    except (json.JSONDecodeError, ValueError):
        parsed = text

    return PublishResult(
        status=status,
        ok=200 <= status < 300,
        body=parsed,
        name=str(manifest["name"]),
        version=str(manifest["version"]),
        artifact_type=resolved_type,
    )

File: packages/rust/alex-sdk/src/publish.rs
//! Publish a packed `.atool` / `.aagent` archive to an Alexandria registry.
//!
//! Mirrors the TypeScript SDK `publish()`: the archive is re-verified (hashes +
//! schema) then a multipart body is POSTed to `{registry}/v1/submit` — the
//! missing consumer-side half of the registry loop (`alexandria install` pulls;
//! nothing pushed until now).
//!
//! The HTTP send is abstracted behind the [`Transport`] trait so callers (and
//! tests) can inject a mock. [`publish`] uses the built-in [`UreqTransport`].

use std::path::Path;

use crate::pack::verify;
use crate::{Error, Result};

/// Outcome of a publish attempt.
#[derive(Debug, Clone)]
pub struct PublishResult {
    pub status: u16,
    pub ok: bool,
    pub body: String,
    pub name: String,
    pub version: String,
    pub artifact_type: String,
}

/// Options controlling a publish call.
#[derive(Debug, Clone, Default)]
pub struct PublishOptions {
    /// Optional bearer token (sent as `Authorization: Bearer <token>`).
    pub token: Option<String>,
    /// Override the artifact_type; defaults to the manifest's kind.
    pub artifact_type: Option<String>,
}

/// Pluggable HTTP transport. `send` receives the fully-formed URL, the raw
/// multipart body, its content-type, and an optional bearer token, and returns
/// `(status, response_body)`.
pub trait Transport {
    fn send(
        &self,
        url: &str,
        content_type: &str,
        body: &[u8],
        token: Option<&str>,
    ) -> Result<(u16, String)>;
}

/// Default transport backed by the `ureq` blocking HTTP client.
pub struct UreqTransport;

impl Transport for UreqTransport {
    fn send(
        &self,
        url: &str,
        content_type: &str,
        body: &[u8],
        token: Option<&str>,
    ) -> Result<(u16, String)> {
        let mut req = ureq::post(url).header("content-type", content_type);
        if let Some(t) = token {
            req = req.header("authorization", &format!("Bearer {t}"));
        }
        match req.send(body) {
            Ok(mut resp) => {
                let status = resp.status().as_u16();
                let text = resp.body_mut().read_to_string().unwrap_or_default();
                Ok((status, text))
            }
            // ureq returns Err for non-2xx status codes; surface them as a
            // normal (status, body) pair rather than a hard error.
            Err(ureq::Error::StatusCode(code)) => Ok((code, String::new())),
            Err(e) => Err(Error::Other(format!("http transport error: {e}"))),
        }
    }
}

/// Assemble a `multipart/form-data` body carrying an `artifact_type` text field
/// and a `tarball` file part. Returns `(body, content_type)`.
pub fn build_multipart(artifact_type: &str, filename: &str, tarball: &[u8]) -> (Vec<u8>, String) {
    let boundary = format!("----alexsdk{:x}", fastrand_boundary());
    let mut body: Vec<u8> = Vec::with_capacity(tarball.len() + 256);

    body.extend_from_slice(format!("--{boundary}\r\n").as_bytes());
    body.extend_from_slice(b"Content-Disposition: form-data; name=\"artifact_type\"\r\n\r\n");
    body.extend_from_slice(artifact_type.as_bytes());
    body.extend_from_slice(b"\r\n");

    body.extend_from_slice(format!("--{boundary}\r\n").as_bytes());
    body.extend_from_slice(
        format!("Content-Disposition: form-data; name=\"tarball\"; filename=\"{filename}\"\r\n")
            .as_bytes(),
    );
    body.extend_from_slice(b"Content-Type: application/gzip\r\n\r\n");
    body.extend_from_slice(tarball);
    body.extend_from_slice(b"\r\n");

    body.extend_from_slice(format!("--{boundary}--\r\n").as_bytes());

    let content_type = format!("multipart/form-data; boundary={boundary}");
    (body, content_type)
}

/// Cheap boundary nonce — no crypto needed, just uniqueness within a process.
fn fastrand_boundary() -> u128 {
    use std::time::{SystemTime, UNIX_EPOCH};
    let nanos = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_nanos())
        .unwrap_or(0);
    nanos ^ (std::process::id() as u128).rotate_left(64)
}

/// Publish `pkg_path` to `{registry}/v1/submit` using the given transport.
///
/// The archive is re-verified before it ships — never publish an archive the
/// local runtime would itself reject.
pub fn publish_with<T: Transport>(
    pkg_path: &Path,
    registry: &str,
    opts: &PublishOptions,
    transport: &T,
) -> Result<PublishResult> {
    let manifest = verify(pkg_path)?;
    let kind = serde_json::to_value(&manifest.kind)
        .ok()
        .and_then(|v| v.as_str().map(|s| s.to_string()))
        .unwrap_or_default();
    let artifact_type = opts.artifact_type.clone().unwrap_or(kind);

    let tarball = std::fs::read(pkg_path)?;
    let filename = pkg_path
        .file_name()
        .map(|s| s.to_string_lossy().into_owned())
        .unwrap_or_else(|| "package".to_string());
    let (body, content_type) = build_multipart(&artifact_type, &filename, &tarball);

    let base = registry.trim_end_matches('/');
    let url = format!("{base}/v1/submit");

    let (status, text) = transport.send(&url, &content_type, &body, opts.token.as_deref())?;

    Ok(PublishResult {
        status,
        ok: (200..300).contains(&status),
        body: text,
        name: manifest.name,
        version: manifest.version,
        artifact_type,
    })
}

/// Publish `pkg_path` using the default [`UreqTransport`].
pub fn publish(pkg_path: &Path, registry: &str, opts: &PublishOptions) -> Result<PublishResult> {
    publish_with(pkg_path, registry, opts, &UreqTransport)
}
